statsmodels/robust/covariance.py

```diff
@@ -0,0 +1,910 @@
+        specified the denominator for the weighted covariance
+        If None, then the sum of weights are used and the covariance is
+        an average cross product
+        If "det", then the weighted covariance is normalized such that
+        det(wcov) is 1.
+        If float, then it is used directly as denominator.
+    ddof : int or float
+        covariance degrees of freedom correction, only used if
+        weights_cov_denom is None or a float.
+
+    Notes
+    -----
+    The extra options are available to cover the general M-estimator
+    for location and scatter with estimating equations (using data x):
+
+    sum (weights * (x - m)) = 0
+    sum (weights_cov * (x_i - m) * (x_i - m)') - weights_cov_denom * cov = 0
+
+    where the weights are functions of the mahalonibis distance of the
+    residuals, and m is the mean.
+
+    In the default case
+    wmean = ave (w_i x)i)
+    wcov = ave (w_i (x_i - m) (x_i - m)')
+
+    References
+    ----------
+    .. [1] Rocke, D. M., and D. L. Woodruff. 1993. Computation of Robust
+       Estimates of Multivariate Location and Shape.
+       Statistica Neerlandica 47 (1): 27-42.
+       doi:10.1111/j.1467-9574.1993.tb01404.x.
+
+
+    """
+
+    wsum = weights.sum()
+    if weights_cov is None:
+        weights_cov = weights
+        wsum_cov = wsum
+    else:
+        wsum_cov = None  # calculate below only if needed
+
+    if center is None:
+        wmean = weights.dot(data) / wsum
+    else:
+        wmean = center
+
+    xdm = data - wmean
+    wcov = (weights_cov * xdm.T).dot(xdm)
+    if weights_cov_denom is None:
+        if wsum_cov is None:
+            wsum_cov = weights_cov.sum()
+        wcov /= (wsum_cov - ddof)
+    elif weights_cov_denom == "det":
+        wcov /= np.linalg.det(wcov)**(1 / wcov.shape[0])
+    elif weights_cov_denom == 1:
+        pass
+    else:
+        wcov /= (weights_cov_denom - ddof)
+
+    return wcov, wmean
+
+
+def weights_mvt(distance, df, k_vars):
+    """weight function based on multivariate t distribution
+
+    Parameters
+    ----------
+    distance : ndarray
+        mahalanobis distance
+    df : int or float
+        degrees of freedom of the t distribution
+    k_vars : int
+        number of variables in the multivariate sample
+
+    Returns
+    -------
+    weights : ndarray
+        weights calculated for the given distances.
+
+    References
+    ----------
+    .. [1] Finegold, Michael A., and Mathias Drton. 2014. Robust Graphical
+       Modeling with T-Distributions. arXiv:1408.2033 [Cs, Stat], August.
+       http://arxiv.org/abs/1408.2033.
+
+    .. [2] Finegold, Michael, and Mathias Drton. 2011. Robust graphical
+       modeling of gene networks using classical and alternative
+       t-distributions. The Annals of Applied Statistics 5 (2A): 1057-80.
+    """
+    w = (df + k_vars) / (df + distance)
+    return w
+
+
+def weights_quantile(distance, frac=0.5, rescale=True):
+    """Weight function for cutoff weights.
```

Comment:
this uses fractional trimming.
maybe add option for threshold/metric trimming, i.e. keyword `cutoff` that overrides frac.